benchmarks/imagenet/resnet50/mocov2.py

```diff
@@ -0,0 +1,145 @@
+        x, shuffle = batch_shuffle(batch=x, distributed=self.trainer.num_devices > 1)
+        features = self.forward(x).flatten(start_dim=1)
+        projections = self.projection_head(features)
+        features = batch_unshuffle(
+            batch=features,
+            shuffle=shuffle,
+            distributed=self.trainer.num_devices > 1,
+        )
+        projections = batch_unshuffle(
+            batch=projections,
+            shuffle=shuffle,
+            distributed=self.trainer.num_devices > 1,
+        )
+        return features, projections
+
+    def forward_query_encoder(self, x: Tensor) -> Tensor:
+        features = self.query_backbone(x).flatten(start_dim=1)
+        projections = self.query_projection_head(features)
+        return projections
+
+    def training_step(
+        self, batch: Tuple[List[Tensor], Tensor, List[str]], batch_idx: int
+    ) -> Tensor:
+        views, targets = batch[0], batch[1]
+
+        # Encode queries.
+        query_projections = self.forward_query_encoder(views[1])
+
+        # Momentum update. This happens between query and key encoding, following the
+        # original implementation from the authors:
+        # https://github.com/facebookresearch/moco/blob/5a429c00bb6d4efdf511bf31b6f01e064bf929ab/moco/builder.py#L142
+        update_momentum(self.query_backbone, self.backbone, m=0.999)
+        update_momentum(self.query_projection_head, self.projection_head, m=0.999)
+
+        # Encode keys.
+        key_features, key_projections = self.forward_key_encoder(views[0])
+        loss = self.criterion(query_projections, key_projections)
+        self.log(
+            "train_loss", loss, prog_bar=True, sync_dist=True, batch_size=len(targets)
+        )
+
+        # Online linear evaluation.
+        cls_loss, cls_log = self.online_classifier.training_step(
+            (key_features.detach(), targets), batch_idx
+        )
+        self.log_dict(cls_log, sync_dist=True, batch_size=len(targets))
+        return loss + cls_loss
+
+    def validation_step(
+        self, batch: Tuple[Tensor, Tensor, List[str]], batch_idx: int
+    ) -> Tensor:
+        images, targets = batch[0], batch[1]
+        features = self.forward(images).flatten(start_dim=1)
+        cls_loss, cls_log = self.online_classifier.validation_step(
+            (features.detach(), targets), batch_idx
+        )
+        self.log_dict(cls_log, prog_bar=True, sync_dist=True, batch_size=len(targets))
+        return cls_loss
+
+    def configure_optimizers(self):
+        # Don't use weight decay for batch norm, bias parameters, and classification
+        # head to improve performance.
+        # NOTE: The original implementation from the authors uses weight decay for all
+        # parameters.
+        params, params_no_weight_decay = get_weight_decay_parameters(
+            [self.query_backbone, self.query_projection_head]
+        )
+        optimizer = SGD(
```

Comment:
was SGD also used in the paper? I was a bit surprised by the choice of optimizer

Reply:
Yes :) SGD with momentum is still a popular choice in self-supervised learning.